private/bufpkg/bufimage/bufimageutil/bufimageutil.go

```diff
@@ -60,6 +61,470 @@ func FreeMessageRangeStrings(
+	for _, typeName := range types {
+		descriptor, ok := imageIndex.NameToDescriptor[typeName]
+		if !ok {
+			return nil, fmt.Errorf("not found: %q", typeName)
+		}
+		if image.GetFile(descriptor.File().Path()).IsImport() {
+			return nil, fmt.Errorf("type %q is in an import", typeName)
+		}
+		startingDescriptors = append(startingDescriptors, descriptor)
+	}
+	// Find all types to include in filtered image.
+	seen := make(map[string]struct{})
+	neededDescriptors := []descriptorAndDirects{}
+	for _, startingDescriptor := range startingDescriptors {
+		closure, err := descriptorTransitiveClosure(startingDescriptor, imageIndex, seen)
+		if err != nil {
+			return nil, err
+		}
+		neededDescriptors = append(neededDescriptors, closure...)
+	}
+	// By file is more convenient but we loose dfs order from
+	// descriptorTransitiveClosure, if images need files ordered by
+	// dependencies may need to rethink this.
+	descriptorsByFile := make(map[string][]descriptorAndDirects)
+	for _, descriptor := range neededDescriptors {
+		descriptorsByFile[descriptor.Descriptor.File().Path()] = append(
+			descriptorsByFile[descriptor.Descriptor.File().Path()],
+			descriptor,
+		)
+	}
+	// Create a new image with only the required descriptors.
+	var includedFiles []bufimage.ImageFile
+	for file, descriptors := range descriptorsByFile {
+		importsRequired := make(map[string]struct{})
+		typesToKeep := make(map[string]struct{})
+		for _, descriptor := range descriptors {
+			typesToKeep[descriptor.Descriptor.FullName()] = struct{}{}
+			for _, importedDescdescriptor := range descriptor.Directs {
+				if importedDescdescriptor.File() != descriptor.Descriptor.File() {
+					importsRequired[importedDescdescriptor.File().Path()] = struct{}{}
+				}
+			}
+		}
+
+		imageFile := image.GetFile(file)
+		includedFiles = append(includedFiles, imageFile)
+		imageFileDescriptor := imageFile.Proto()
+		// While employing
+		// https://github.com/golang/go/wiki/SliceTricks#filter-in-place,
+		// also keep a record of which index moved where so we can fixup
+		// the file's PublicDependency/WeakDependency fields.
+		indexFromTo := make(map[int32]int32)
+		indexTo := 0
+		for indexFrom, importPath := range imageFileDescriptor.GetDependency() {
+			// TODO: this only filters the existing imports down to
+			// the ones requested, if there was a type we picked up
+			// through a public import in a dependent file may
+			// filter out that file here, a type not to be found. We
+			// may need to add the file directly (or have a file
+			// with public import only inserted in the middle).
+			// We should check if all keys in importmap get looked up.
+			if _, ok := importsRequired[importPath]; ok {
+				indexFromTo[int32(indexFrom)] = int32(indexTo)
+				imageFileDescriptor.Dependency[indexTo] = importPath
+				indexTo++
+			}
+		}
+		imageFileDescriptor.Dependency = imageFileDescriptor.Dependency[:indexTo]
+		i := 0
```

Comment:
Use `var i int` for default value initialization whenever possible. Applies elsewhere.